### ISSUE TITLE
`@deque.T`'s implementation is not correct

### ISSUE THREAD

**Issue by rami3l** [moonbitlang/core#1361]
Minimal reproduction below:

```moonbit
fn main {
  let d = @deque.of(["RP", "NL"])
  ignore(d.pop_front())
  guard let Some(f) = d.pop_front()
  d.push_back(f)
  println("\{d}") // Expected `@deque.of(["NL"])`; got `@deque.of(["RP"])`.
}
```

**Comment by rami3l** [comment]
I suggest a rewrite of this module since `tail` can be computed and should not be a saved state.

**Comment by peter-jerry-ye** [comment]
The invariant of deque is broken: #1365

**Comment by rami3l** [comment]
Closed by #1374.
